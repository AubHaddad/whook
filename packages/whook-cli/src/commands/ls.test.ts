import initLsCommand, { definition as initLsCommandDefinition } from './ls';
import initEnvCommand, { definition as initEnvCommandDefinition } from './env';
import YError from 'yerror';

describe('lsCommand', () => {
  const promptArgs = jest.fn();
  const log = jest.fn();
  const readDir = jest.fn();
  const importer = jest.fn();

  beforeEach(() => {
    promptArgs.mockReset();
    log.mockReset();
    readDir.mockReset();
    importer.mockReset();
  });

  describe('should work', () => {
    it('with no plugin', async () => {
      promptArgs.mockResolvedValueOnce({
        _: ['ls'],
      });
      readDir.mockRejectedValueOnce(new YError('E_NO_MODULE'));

      const lsCommand = await initLsCommand({
        CONFIG: {
          name: 'My Super project!',
        },
        PROJECT_SRC: '/home/whoiam/whook-project/dist',
        WHOOK_PLUGINS: [],
        WHOOK_PLUGINS_PATHS: [],
        promptArgs,
        readDir,
        log,
        EOL: '\n',
        importer,
      });
      const result = await lsCommand();

      expect({
        result,
        promptArgsCalls: promptArgs.mock.calls,
        logCalls: log.mock.calls.filter(([type]) => !type.endsWith('stack')),
        readDirCalls: readDir.mock.calls,
        requireCalls: importer.mock.calls,
      }).toMatchSnapshot();
    });

    it('with some plugins', async () => {
      readDir.mockResolvedValueOnce(['ls', 'env']);
      readDir.mockRejectedValueOnce(new YError('E_NO_MODULE'));
      importer.mockReturnValueOnce({
        default: initLsCommand,
        definition: initLsCommandDefinition,
      });
      importer.mockReturnValueOnce({
        default: initEnvCommand,
        definition: initEnvCommandDefinition,
      });
      promptArgs.mockResolvedValueOnce({
        _: ['ls'],
      });

      const lsCommand = await initLsCommand({
        CONFIG: {},
        PROJECT_SRC: '/home/whoiam/whook-project/dist',
        WHOOK_PLUGINS: ['@whook/cli', '@whook/whook'],
        WHOOK_PLUGINS_PATHS: [
          '/var/lib/node/node_modules/@whook/cli/dist',
          '/var/lib/node/node_modules/@whook/lol/dist',
        ],
        promptArgs,
        readDir,
        log,
        EOL: '\n',
        importer,
      });
      await lsCommand();

      expect({
        promptArgsCalls: promptArgs.mock.calls,
        logCalls: log.mock.calls.filter(([type]) => !type.endsWith('stack')),
        readDirCalls: readDir.mock.calls,
        requireCalls: importer.mock.calls,
      }).toMatchSnapshot();
    });

    it('with some plugins and ignored files', async () => {
      readDir.mockResolvedValueOnce(['ls', 'env', '__snapshots__']);
      readDir.mockRejectedValueOnce(new YError('E_NO_MODULE'));
      importer.mockReturnValueOnce({
        default: initLsCommand,
        definition: initLsCommandDefinition,
      });
      importer.mockReturnValueOnce({
        default: initEnvCommand,
        definition: initEnvCommandDefinition,
      });
      promptArgs.mockResolvedValueOnce({
        _: ['ls'],
      });

      const lsCommand = await initLsCommand({
        CONFIG: {},
        PROJECT_SRC: '/home/whoiam/whook-project/dist',
        WHOOK_PLUGINS: ['@whook/cli', '@whook/whook'],
        WHOOK_PLUGINS_PATHS: [
          '/var/lib/node/node_modules/@whook/cli/dist',
          '/var/lib/node/node_modules/@whook/lol/dist',
        ],
        promptArgs,
        readDir,
        log,
        EOL: '\n',
        importer,
      });
      await lsCommand();

      expect({
        promptArgsCalls: promptArgs.mock.calls,
        logCalls: log.mock.calls.filter(([type]) => !type.endsWith('stack')),
        readDirCalls: readDir.mock.calls,
        requireCalls: importer.mock.calls,
      }).toMatchSnapshot();
    });

    it('with some plugins and a verbose output', async () => {
      readDir.mockResolvedValueOnce(['ls', 'env']);
      readDir.mockRejectedValueOnce(new YError('E_NO_MODULE'));
      importer.mockReturnValueOnce({
        default: initLsCommand,
        definition: initLsCommandDefinition,
      });
      importer.mockReturnValueOnce({
        default: initEnvCommand,
        definition: initEnvCommandDefinition,
      });
      promptArgs.mockResolvedValueOnce({
        _: ['ls'],
        verbose: true,
      });

      const lsCommand = await initLsCommand({
        CONFIG: {
          name: 'My Super project!',
        },
        PROJECT_SRC: '/home/whoiam/whook-project/dist',
        WHOOK_PLUGINS: ['@whook/cli', '@whook/whook'],
        WHOOK_PLUGINS_PATHS: [
          '/var/lib/node/node_modules/@whook/cli/dist',
          '/var/lib/node/node_modules/@whook/lol/dist',
        ],
        promptArgs,
        readDir,
        log,
        EOL: '\n',
        importer,
      });
      await lsCommand();

      expect({
        promptArgsCalls: promptArgs.mock.calls,
        logCalls: log.mock.calls.filter(([type]) => !type.endsWith('stack')),
        readDirCalls: readDir.mock.calls,
        requireCalls: importer.mock.calls,
      }).toMatchSnapshot();
    });
  });
});
