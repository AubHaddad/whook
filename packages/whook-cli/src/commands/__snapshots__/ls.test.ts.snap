// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lsCommand should work with no plugin 1`] = `
{
  "logCalls": [
    [
      "debug",
      "✅ - No commands folder found at path "/home/whoiam/whook-project/dist".",
    ],
    [
      "info",
      "

# Provided by "My Super project!": none",
    ],
  ],
  "promptArgsCalls": [
    [],
  ],
  "readDirCalls": [
    [
      "/home/whoiam/whook-project/dist/commands",
    ],
  ],
  "requireCalls": [],
  "result": undefined,
}
`;

exports[`lsCommand should work with some plugins 1`] = `
{
  "logCalls": [
    [
      "debug",
      "✅ - No commands folder found at path "/var/lib/node/node_modules/@whook/lol/dist".",
    ],
    [
      "debug",
      "✅ - No commands folder found at path "/var/lib/node/node_modules/@whook/cli/dist".",
    ],
    [
      "info",
      "

# Provided by "project": 2 commands",
    ],
    [
      "info",
      "- ls: Print available commands",
    ],
    [
      "info",
      "- env: A command printing env values",
    ],
    [
      "info",
      "

# Provided by "@whook/cli": none",
    ],
    [
      "info",
      "

# Provided by "@whook/whook": none",
    ],
  ],
  "promptArgsCalls": [
    [],
  ],
  "readDirCalls": [
    [
      "/home/whoiam/whook-project/dist/commands",
    ],
    [
      "/var/lib/node/node_modules/@whook/cli/dist/commands",
    ],
    [
      "/var/lib/node/node_modules/@whook/lol/dist/commands",
    ],
  ],
  "requireCalls": [
    [
      "/home/whoiam/whook-project/dist/commands/ls.js",
    ],
    [
      "/home/whoiam/whook-project/dist/commands/env.js",
    ],
  ],
}
`;

exports[`lsCommand should work with some plugins and a verbose output 1`] = `
{
  "logCalls": [
    [
      "debug",
      "✅ - No commands folder found at path "/var/lib/node/node_modules/@whook/lol/dist".",
    ],
    [
      "debug",
      "✅ - No commands folder found at path "/var/lib/node/node_modules/@whook/cli/dist".",
    ],
    [
      "info",
      "

# Provided by "My Super project!": 2 commands
",
    ],
    [
      "info",
      "- ls: Print available commands
$ whook ls
",
    ],
    [
      "info",
      "- env: A command printing env values
$ whook env --name NODE_ENV --default "default value"
",
    ],
    [
      "info",
      "

# Provided by "@whook/cli": none
",
    ],
    [
      "info",
      "

# Provided by "@whook/whook": none
",
    ],
  ],
  "promptArgsCalls": [
    [],
  ],
  "readDirCalls": [
    [
      "/home/whoiam/whook-project/dist/commands",
    ],
    [
      "/var/lib/node/node_modules/@whook/cli/dist/commands",
    ],
    [
      "/var/lib/node/node_modules/@whook/lol/dist/commands",
    ],
  ],
  "requireCalls": [
    [
      "/home/whoiam/whook-project/dist/commands/ls.js",
    ],
    [
      "/home/whoiam/whook-project/dist/commands/env.js",
    ],
  ],
}
`;

exports[`lsCommand should work with some plugins and ignored files 1`] = `
{
  "logCalls": [
    [
      "debug",
      "✅ - No commands folder found at path "/var/lib/node/node_modules/@whook/lol/dist".",
    ],
    [
      "debug",
      "✅ - No commands folder found at path "/var/lib/node/node_modules/@whook/cli/dist".",
    ],
    [
      "info",
      "

# Provided by "project": 2 commands",
    ],
    [
      "info",
      "- ls: Print available commands",
    ],
    [
      "info",
      "- env: A command printing env values",
    ],
    [
      "info",
      "

# Provided by "@whook/cli": none",
    ],
    [
      "info",
      "

# Provided by "@whook/whook": none",
    ],
  ],
  "promptArgsCalls": [
    [],
  ],
  "readDirCalls": [
    [
      "/home/whoiam/whook-project/dist/commands",
    ],
    [
      "/var/lib/node/node_modules/@whook/cli/dist/commands",
    ],
    [
      "/var/lib/node/node_modules/@whook/lol/dist/commands",
    ],
  ],
  "requireCalls": [
    [
      "/home/whoiam/whook-project/dist/commands/ls.js",
    ],
    [
      "/home/whoiam/whook-project/dist/commands/env.js",
    ],
  ],
}
`;
